misc/mcnaughton_yamada_thompson.c

```diff
@@ -0,0 +1,719 @@
+            // Gradually empties the current states pool, so
+            // it can be refilled
+            nfa->CSCount--;
+
+            // Iterates through rules of this state
+            for (int j = 0; j < pState->ruleCount; ++j) {
+                const struct transRule *pRule = pState->rules[j];
+
+                if(pRule->cond == input) {
+                    if(!contains(newStates, NSCount, pRule->target)) {
+                        newStates[NSCount++] = pRule->target;
+                    }
+                }
+            }
+        }
+    }
+
+    nfa->CSCount = NSCount;
+    for (int i = 0; i < NSCount; ++i) {
+        nfa->currentStates[i] = newStates[i];
+    }
+    free(newStates);
+}
+
+/**
+ * @brief determines whether the NFA is currently in its accepting state
+ * @param nfa target NFA
+ * @returns `1` if the NFA is in its accepting state
+ * @returns `0` otherwise
+ */
+int isAccepting(const struct NFA* nfa) {
+    for (int i = 0; i < nfa->CSCount; ++i) {
+        if(nfa->currentStates[i] == nfa->statePool[1]) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+/* Ends NFA utilities, begins testing function*/
+
+/**
+ * @brief Testing helper function
+ * @param regex the regular expression to be used
+ * @param string the string to match against
+ * @param expected expected results
+ * @returns void
+ */
+void testHelper(const char* regex, const char* string, const int expected) {
+    char* temp = preProcessing(regex);
+    struct ASTNode* node = buildAST(temp);
+
+    struct NFA* nfa = compileFromAST(node);
+    postProcessing(nfa);
+
+    // reallocates the outermost NFA's current states pool
+    // because it will actually be used to store all the states
+    nfa->currentStates = realloc(nfa->currentStates, sizeof(struct NFAState*) * 100);
+    // Starts the NFA by adding its starting state to the pool
+    nfa->currentStates[nfa->CSCount++] = nfa->statePool[0];
+
+    // feeds empty characters into the NFA before and after
+    // every normal character
+    for (size_t i = 0; i < strlen(string); ++i) {
+        transit(nfa, '\0');
+        transit(nfa, string[i]);
+    }
+    transit(nfa, '\0');
+
+    assert(isAccepting(nfa) == expected);
+
+    destroyNFA(nfa);
+    destroyNode(node);
+    free(temp);
+}
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test(void) {
+    testHelper("(c|a*b)", "c", 1);
+    testHelper("(c|a*b)", "aab", 1);
+    testHelper("(c|a*b)", "ca", 0);
+    testHelper("(c|a*b)*", "caaab", 1);
+    testHelper("(c|a*b)*", "caba", 0);
+    testHelper("", "", 1);
+    testHelper("", "1", 0);
+    testHelper("(0|(1(01*(00)*0)*1)*)*","11",1);
+    testHelper("(0|(1(01*(00)*0)*1)*)*","110",1);
+    testHelper("(0|(1(01*(00)*0)*1)*)*","1100",1);
+    testHelper("(0|(1(01*(00)*0)*1)*)*","10000",0);
+    testHelper("(0|(1(01*(00)*0)*1)*)*","00000",1);
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main(void) {
```

Comment:
Unnecessary, IMO.
```suggestion
int main() {
```

Reply:
Actually, I specifically research this some time ago, and it turns out there is a difference between `int main()` and `int main(void)`. `int main()` accepts an unspecified amount of arguments while `int main(void)` explicitly accepts no arguments. If you crank the warning level high enough, your compiler / IDE will complain about `function declaration without prototype is deprecated in all versions of C`. [Source](https://stackoverflow.com/questions/42125/warning-error-function-declaration-isnt-a-prototype)

Reply:
BTW, I code with this much warning on: `add_compile_options(-Wall -Wextra -Wpedantic -Wformat=2 -Wno-unused-parameter
        -Wshadow -Wwrite-strings -Wstrict-prototypes -Wold-style-definition
        -Wredundant-decls -Wnested-externs -Wmissing-include-dirs -Wjump-misses-init -Wlogical-op)`